connector/connect/src/main/scala/org/apache/spark/sql/connect/service/SparkConnectStreamHandler.scala

```diff
@@ -117,10 +129,91 @@ class SparkConnectStreamHandler(responseObserver: StreamObserver[Response]) exte
       responseObserver.onNext(response.build())
     }
 
-    responseObserver.onNext(sendMetricsToResponse(clientId, rows))
+    responseObserver.onNext(sendMetricsToResponse(clientId, dataframe))
     responseObserver.onCompleted()
   }
 
+  def processRowsAsArrowBatches(clientId: String, dataframe: DataFrame): Unit = {
+    val spark = dataframe.sparkSession
+    val schema = dataframe.schema
+    // TODO: control the batch size instead of max records
+    val maxRecordsPerBatch = spark.sessionState.conf.arrowMaxRecordsPerBatch
+    val timeZoneId = spark.sessionState.conf.sessionLocalTimeZone
+
+    SQLExecution.withNewExecutionId(dataframe.queryExecution, Some("collectArrow")) {
+      val pool = ThreadUtils.newDaemonSingleThreadExecutor("connect-collect-arrow")
+      val tasks = collection.mutable.ArrayBuffer.empty[Future[_]]
+      val rows = dataframe.queryExecution.executedPlan.execute()
+
+      if (rows.getNumPartitions > 0) {
+        val batches = rows.mapPartitionsInternal { iter =>
+          ArrowConverters
+            .toArrowBatchIterator(iter, schema, maxRecordsPerBatch, timeZoneId)
+        }
+
+        val processPartition = (iter: Iterator[(Array[Byte], Long, Long)]) => iter.toArray
+
+        val resultHandler = (partitionId: Int, taskResult: Array[(Array[Byte], Long, Long)]) => {
+          if (taskResult.exists(_._1.nonEmpty)) {
+            // only send non-empty partitions
+            val task = pool.submit(new Runnable {
+              override def run(): Unit = {
+                var batchId = partitionId.toLong << 33
+                taskResult.foreach { case (bytes, count, size) =>
+                  val response = proto.Response.newBuilder().setClientId(clientId)
+                  val batch = proto.Response.ArrowBatch
+                    .newBuilder()
+                    .setBatchId(batchId)
+                    .setRowCount(count)
+                    .setUncompressedBytes(size)
+                    .setCompressedBytes(bytes.length)
+                    .setData(ByteString.copyFrom(bytes))
+                    .build()
+                  response.setArrowBatch(batch)
+                  responseObserver.onNext(response.build())
```

Comment:
I feel like this comes with a bigger problem to define the client side incremental collect protocol first. Client side consumes data but does not consume them all at once. How client incremental collect? Then combined with that how server side control the rate.

Reply:
Things goes worse if the we decide to keep strict partition ordering. 

Actually, Spark implemented `IndirectTaskResult` to support transfer task result from executor to driver through block manager, but currently it eagerly fetches and deserializes blocks to `DirectTaskResult` on `task-result-getter` thread pool. What if defer it to `JobWaiter#taskSucceeded`?

Reply:
You could do that yes. If we get the format right we could just serve the block directly to the client (to avoid deserialization).

Alternatively you could also write the data to some persistent storage on the executors, and just pass the file paths to the client. That will be the most efficient from the POV of the driver. This requires you have some garbage collection in place to clean-up results.

Reply:
To be clear, my suggestion is to defer the deserialization and block releasing in `JobWaiter#taskSucceeded` phase.

The second way you suggested has many performance benefits but requires the client to communicate with the external storage service.

The `IndirectTaskResult` way leverages the existing code and spark build-in block mechanism to transfer data, we can benefit w/ a little code modification, and we don't need to worry about result cleanup.